### PR TITLE
BUG `DataFrameGroupBy.agg` with list not respecting `as_index=False`

### LINKED ISSUE #52849: BUG: DataFrameGroupBy.agg with lists doesn't respect as_index=False
```
df = DataFrame({"a1": [0, 0, 1], "a2": [2, 3, 3], "b": [4, 5, 6]})
# df = df.astype({"a1": "category", "a2": "category"})
gb = df.groupby(by=["a1", "a2"], as_index=False)
result = gb.agg(["sum"])
print(result)
#         b
#       sum
# a1 a2
# 0  2    4
#    3    5
# 1  3    6
```

The expected output (which you get with `gb.agg("sum")`) is:

```
#    a1  a2  b
# 0   0   2  4
# 1   0   3  5
# 2   1   3  6
```

### PR DESCRIPTION
**Pull request by Charlie-XIAO**
- [x] Closes #52849(
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit)
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints)
- [x] Added an entry in the latest `doc/source/whatsnew/v2.1.0.rst` file

Dealing with non-categorical data is fairly easy, but dealing with categorical data is kind of tricky. I'm currently taking a pretty dirty approach, so if maintainers have better solutions please let me know and I'll make changes correspondingly.

### CODE REVIEW

**Comment by Charlie-XIAO** [comment]
Sorry I think I forgot to clean up the code. I will do that ASAP.

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -1396,9 +1396,7 @@ def aggregate(self, func=None, *args, engine=None, engine_kwargs=None, **kwargs)
 
         op = GroupByApply(self, func, args=args, kwargs=kwargs)
         result = op.agg()
-        if not is_dict_like(func) and result is not None:
-            return result
```
Comment (rhshadrach) [inline pandas/core/groupby/generic.py:1400]: Can we just call `.reset_index()` when `as_index=False` here?

Comment (Charlie-XIAO) [reply pandas/core/groupby/generic.py:1400]: I think you are right @rhshadrach, this seems to be the right way to go. However, simply doing this is failing certain test cases. I will try to investigate later.

Comment (Charlie-XIAO) [reply pandas/core/groupby/generic.py:1400]: Update: according to my investigation, it needs to (at least) be something like
```python
return result if self.as_index or not is_list_like(func) else result.reset_index()
```
I will modify the PR in this way for now.

File: pandas/core/groupby/generic.py [hunk 2]
```diff
@@ -1397,7 +1398,11 @@ def aggregate(self, func=None, *args, engine=None, engine_kwargs=None, **kwargs)
         op = GroupByApply(self, func, args=args, kwargs=kwargs)
         result = op.agg()
         if not is_dict_like(func) and result is not None:
-            return result
+            # GH #52849
+            if is_list_like(func) and not self.as_index:
```
Comment (rhshadrach) [inline pandas/core/groupby/generic.py:1402]: Can you switch the order of these two checks; best to have the fastest check first for short-circuiting.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Can you add tests for this other than categorical in `tests/groupby/aggregate/test_aggregate.py`

**Comment by Charlie-XIAO** [comment]
@rhshadrach Thanks for your review! New test added and I've made your suggested modification.

**Review by rhshadrach — APPROVED** [review]
lgtm

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @Charlie-XIAO

**Pull request merged**